process/smartContract/processorV2/vmInputV2.go

```diff
@@ -0,0 +1,204 @@
+	}
+
+	vmCreateInput.VMInput.Arguments = deployData.Arguments
+
+	return vmCreateInput, deployData.VMType, nil
+}
+
+func (sc *scProcessor) initializeVMInputFromTx(vmInput *vmcommon.VMInput, tx data.TransactionHandler) error {
+	var err error
+
+	vmInput.CallerAddr = tx.GetSndAddr()
+	vmInput.CallValue = new(big.Int).Set(tx.GetValue())
+	vmInput.GasPrice = tx.GetGasPrice()
+	vmInput.GasProvided, err = sc.prepareGasProvided(tx)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func isSmartContractResult(tx data.TransactionHandler) bool {
+	_, isScr := tx.(*smartContractResult.SmartContractResult)
+	return isScr
+}
+
+func (sc *scProcessor) prepareGasProvided(tx data.TransactionHandler) (uint64, error) {
+	if isSmartContractResult(tx) {
+		return tx.GetGasLimit(), nil
+	}
+
+	if sc.shardCoordinator.ComputeId(tx.GetSndAddr()) == core.MetachainShardId {
+		return tx.GetGasLimit(), nil
+	}
+
+	gasForTxData := sc.economicsFee.ComputeGasLimit(tx)
+	if tx.GetGasLimit() < gasForTxData {
+		return 0, process.ErrNotEnoughGas
+	}
+
+	return tx.GetGasLimit() - gasForTxData, nil
+}
+
+func (sc *scProcessor) createVMCallInput(
+	tx data.TransactionHandler,
+	txHash []byte,
+	builtInFuncCall bool,
+) (*vmcommon.ContractCallInput, error) {
+	callType := determineCallType(tx)
+	txData := string(tx.GetData())
+	if !builtInFuncCall {
+		txData = string(prependLegacyCallbackFunctionNameToTxDataIfAsyncCallBack(tx.GetData(), callType))
+	}
+
+	function, arguments, err := sc.argsParser.ParseCallData(txData)
+	if err != nil {
+		return nil, err
+	}
+
+	finalArguments, gasLocked := sc.getAsyncCallGasLockFromTxData(callType, arguments)
+
+	asyncArguments, callArguments := separateAsyncArguments(callType, finalArguments)
+
+	vmCallInput := &vmcommon.ContractCallInput{}
+	vmCallInput.VMInput = vmcommon.VMInput{}
+	vmCallInput.CallType = callType
+	vmCallInput.RecipientAddr = tx.GetRcvAddr()
+	vmCallInput.Function = function
+	vmCallInput.CurrentTxHash = txHash
+	vmCallInput.GasLocked = gasLocked
+
+	scr, isSCR := tx.(*smartContractResult.SmartContractResult)
+	if isSCR {
+		vmCallInput.OriginalTxHash = scr.GetOriginalTxHash()
+		vmCallInput.PrevTxHash = scr.PrevTxHash
+	} else {
+		vmCallInput.OriginalTxHash = txHash
+		vmCallInput.PrevTxHash = txHash
+	}
+
+	vmCallInput.ReturnCallAfterError = isSCR && len(scr.ReturnMessage) > 0
+
+	err = sc.initializeVMInputFromTx(&vmCallInput.VMInput, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	vmCallInput.VMInput.AsyncArguments = buildAsyncArgumentsObject(callType, asyncArguments)
+	vmCallInput.VMInput.Arguments = callArguments
+	if vmCallInput.GasProvided > tx.GetGasLimit() {
+		return nil, process.ErrInvalidVMInputGasComputation
+	}
+
+	vmCallInput.GasProvided, err = core.SafeSubUint64(vmCallInput.GasProvided, gasLocked)
+	if err != nil {
+		return nil, err
+	}
+
+	return vmCallInput, nil
+}
+
+func (sc *scProcessor) getAsyncCallGasLockFromTxData(callType vm.CallType, arguments [][]byte) ([][]byte, uint64) {
```

Comment:
this can be left without the *sc pointer since it is a new file and does not necessarily depend upon the scProcessor struct

Reply:
fixed